stdlib/public/Observation/Sources/Observation/ObservationRegistrar.swift

```diff
@@ -0,0 +1,427 @@
+        keyPath: PartialKeyPath<Subject>, 
+        phase: Phase, 
+        collected: inout KeyPaths<Subject>?
+      ) -> ResumeAction? {
+        switch (self, phase) {
+        case (.transaction(let observedKeyPaths), .willSet):
+          if observedKeyPaths.contains(keyPath) {
+            if var keyPaths = collected {
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            } else {
+              var keyPaths = KeyPaths<Subject>()
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            }
+          }
+          return nil
+        case (.change(let observedKeyPaths, let continuation), .didSet):
+          if observedKeyPaths.contains(keyPath) {
+            if var keyPaths = collected {
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            } else {
+              var keyPaths = KeyPaths<Subject>()
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            }
+            return .resumeAndRemove(continuation)
+          }
+          return nil
+        case (.tracking(let observedKeyPaths, let observer), .willSet):
+          if observedKeyPaths.contains(keyPath) {
+            if var keyPaths = collected {
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            } else {
+              var keyPaths = KeyPaths<Subject>()
+              keyPaths.insert(keyPath)
+              collected = keyPaths
+            }
+            return .informAndRemove(observer)
+          } else {
+            return nil
+          }
+        default:
+          return nil
+        }
+      }
+      
+      func invalidate(
+        keyPaths: KeyPaths<Subject>, 
+        from lookup: inout [PartialKeyPath<Subject>: Set<Int>], 
+        generation: Int
+      ) {
+        for raw in keyPaths.raw {
+          if var members = lookup[raw] {
+            members.remove(generation)
+            if members.isEmpty {
+              lookup.removeValue(forKey: raw)
+            } else {
+              lookup[raw] = members
+            }
+          }
+        }
+      }
+      
+      func remove(
+        from lookup: inout [PartialKeyPath<Subject>: Set<Int>], 
+        generation: Int
+      ) {
+        switch self {
+        case .transaction(let keyPaths):
+          invalidate(keyPaths: keyPaths, from: &lookup, generation: generation)
+        case .change(let keyPaths, _):
+          invalidate(keyPaths: keyPaths, from: &lookup, generation: generation)
+        case .tracking(let keyPaths, _):
+          invalidate(keyPaths: keyPaths, from: &lookup, generation: generation)
+        default:
+          break
+        }
+      }
+      
+      func deinitialize() {
+        switch self {
+        case .change(_, let continuation):
+          continuation.resume(returning: nil)
+        default:
+          break
+        }
+      }
+    }
+    
+    let kind: Kind
+    var collected: KeyPaths<Subject>?
```

Comment:
Nit: This public struct does not appear to have an explicit initializer -- it's a good idea to add one explicitly, even if it matches whatever the compiler synthesizes on its own.)

(Quick, which struct do these belong to? See other nit about presentation of type definitions.)